src/ur/workday.ur

```diff
@@ -0,0 +1,378 @@
+      * allow manager to see the feedbacks they've requested
+      * allow the subordinate to see the feedback they've been assigned
+     *)
+
+
+    val token =
+        toko <- token;
+        case toko of
+            None => error <xml>You must be logged into Workday to use this feature.</xml>
+          | Some tok => return tok
+
+    val prefixCommon = "https://wcpdev-services1.wd101.myworkday.com/ccx/api/v1/janestreet_wcpdev1"
+    val prefixPerformanceEnablement = "https://wcpdev-services1.wd101.myworkday.com/ccx/api/performanceEnablement/v5/janestreet_wcpdev1"
+    datatype service = Common | PerformanceEnablement
+    fun urlOfService service =
+        case service of
+          Common => prefixCommon
+        | PerformanceEnablement => prefixPerformanceEnablement
+
+    fun logged [a] (_ : show a) (t : transaction a) =
+        v <- t;
+        debug ("Workday response: " ^ show v);
+        return v
+
+    fun api service url =
+        let val prefix = urlOfService service in
+          tok <- token;
+          debug ("Workday GET: " ^ prefix ^ url);
+          logged (WorldFfi.get (bless (prefix ^ url)) (WorldFfi.addHeader WorldFfi.emptyHeaders "Authorization" ("Bearer " ^ tok)) False)
+        end
+
+    fun apiOpt service url =
+        let val prefix = urlOfService service in
+          tok <- token;
+          debug ("Workday GET: " ^ prefix ^ url);
+          logged (WorldFfi.getOpt (bless (prefix ^ url)) (WorldFfi.addHeader WorldFfi.emptyHeaders "Authorization" ("Bearer " ^ tok)) False)
+        end
+
+    fun apiPost service url body =
+        let val prefix = urlOfService service in
+          tok <- token;
+          debug ("Workday POST: " ^ prefix ^ url);
+          debug ("Workday body: " ^ body);
+          logged (WorldFfi.post (bless (prefix ^ url)) (WorldFfi.addHeader WorldFfi.emptyHeaders "Authorization" ("Bearer " ^ tok)) (Some "application/json") body)
+        end
+
+    (* NOTE: could we improve performance by using tables instead of lists? do we need to? *)
+    (* NOTE: upstream this to a world library function *)
+    fun depaginate [object] (j : json object) (limit : int) (service : service) (endpoint : string) : transaction (list object) =
+        let fun depaginate' (offset : int) =
+          jsonResponse <- api service (endpoint ^ "?limit=" ^ (show limit) ^ "&offset=" ^ (show offset));
+          let val resp = (fromJson jsonResponse : response (list object))
+          in
+            if offset >= resp.Total
+            then return resp.Data
+            else
+                rest <- depaginate' (offset + limit);
+                return (List.append resp.Data rest)
+          end
+        in
+          depaginate' 0
+        end
+
+    (* Interface to Workday API for making GET and POST requests. *)
+    structure WorkdayApi = struct
+        structure WorkersApi = struct
+            fun get (workerId : wid) : transaction $worker =
+                s <- api Common ("/workers/" ^ workerId);
```

Comment:
Watch out for code-injection attacks!  Escape values before putting them into URLs or any other interpreted format.

Reply:
Good note, fixed but let's keep an eye out for more